### ISSUE TITLE
Allow numerical seconds for durations. 

### ISSUE THREAD

**Issue by liveaspankaj** [adlnet/xAPI-Spec#588]
It might be useful to allow seconds in numerical form. e.g. Instead of "P20S", using "20" might be easier for those generating statement as well as for storage and comparison. 

As ISO8601 has its own benefits, allowing both "P20S" and "20" might be useful.

**Comment by brianjmiller** [comment]
:-1:. If we must remove the requirement for ISO8601 I'd at least like to see this be an integer rather than a string.

**Comment by liveaspankaj** [comment]
If ISO8601 is removed, I agree it could be integer or rather a float. However, I don't think ISO8601 is going away. So, both formats could be allowed. 

It would save people from learning ISO8601 and writing code to convert formats if they don't need to. And would make it easier to adopt xAPI.

**Comment by fugu13** [comment]
I'm :-1: . The burden on programmers to put the number between a P and an S is very very small, and there's no additional representational power.

**Comment by brianjmiller** [comment]
Right, what @fugu13 said. I didn't mean removing ISO8601 as acceptable, I meant removing it as the *only* acceptable format.

**Comment by garemoko** [comment]
I'm with @fugu13 re: writing code to convert formats, for languages that don't have this functionality natively, I'm very much +1 to getting this added to code libraries. In fact this has just been added to the latest release of TinCanJS out today! 

https://github.com/RusticiSoftware/TinCanJS/releases/tag/0.32.0

**Comment by bscSCORM** [comment]
I'm :-1: on allowing stand-alone seconds, but perhaps we should note that it's valid to track: `PT3690S` and not bother converting to `PT1H1M30S`.

**Comment by liveaspankaj** [comment]
If no one likes the idea, I guess we should at-least add the example that @bscSCORM gave on spec.

**Comment by garemoko** [comment]
There's not really an obvious place to put that example as we don't discuss duration outside the table. Would it be satisfactory to add the ```PT3690S``` example to [Appendix A](https://github.com/adlnet/xAPI-Spec/blob/master/xAPI.md#appendix-a-example-statements)?

**Comment by garemoko** [comment]
Should be good to close this now.

**Comment by canweriotnow** [comment]
@fugu13 I seriously hope this is satire:

> The burden on programmers to put the number between a P and an S is very very small, and there's no additional representational power.

If anyone is doing datetime operations by string concatenation, they need their license revoked. Well, in a fantasy world where we certify software engineers the way we do "real" engineers. 

(In case it's not clear, jk)

More seriously, I fail to see the superiority of ISO 8601 durations over, say, integer milliseconds. 32-bit signed integers can represent the duration of time in ms from 1970-01-01T00:00:00Z to 2038-01-19T03:14:07Z just fine. 

Also, other than joda-time, not many common datetime libraries seem to have strong ISO duration support because no one uses the damn things, that I've encountered. I've spent years of my life that I will never get back hacking away at hairy datetime issues and never even **saw** that part of ISO 8601 until this spec.

So, it's in the spec, we'll deal with it as it stands, but it's near the top of my list o' WTFs to address in 2.x (actually, I should make that an actual list and not just something I joke about).

It really smacks of standards-fetishism (of which I am not innocent, but hey) rather than something of actual value.

**Comment by fugu13** [comment]
Not satire, but sarcasm. ISO 8601 duration can capture a representational superset of integer milliseconds -- they can express *every* duration that can be represented by a number of seconds or milliseconds, and extremely common in practice durations that cannot be expressed as a single number of milliseconds because they are context sensitive, such as "one month" or "one year" can also be expressed exactly using ISO 8601 durations.

For example, it is easy to express that a certificate is valid for one month or one year in a human intuitive fashion with ISO 8601 durations, but impossible to do except by an approximation that actual humans have a hard time working with in milliseconds.

ISO 8601 durations aren't very hard to deal with at all, so I'm sure you'll adapt fast. I'm surprised you haven't run into them, as calendaring standards such as iCalendar use them, for exactly the representational power I mentioned above.

It makes no sense to switch to a representation, with strictly less expressive power, when the current representation isn't hard to use, where transforming from the alternative candidate into the current representation is a pure matter of string concatenation and possibly multiplication, depending on seconds/milliseconds.

**Comment by garemoko** [comment]
On this issue, I'm beginning to wonder if in cases where measurement is done in seconds, we should **recommend** that activity providers use the ```PT2456.3456S``` format over converting to minutes, hours and days. Values should only be given in minutes, hours, days etc. where they are actually measured in those terms. 

Thoughts?

**Comment by fugu13** [comment]
Instead of recommending, I think we can accomplish a similar pedagogical goal succinctly and with more variety by showing a short table of some legal ISO 8601 durations and their human translations, where one of the examples is a large number of seconds.

**Comment by garemoko** [comment]
Perhaps with column describing the scenarios in which each value might be used?

**Comment by DavidTPate** [comment]
I think that the ISO 8601 specification for durations being required is unnecessary and should instead be able to represent the duration as the number of milli-seconds instead of having to use the ISO 8601 duration specification.

The reason for this is because ISO 8601 durations is not a standard that I've found to be supported in most databases and something that in order to query off of would need to be parsed which is unnecessary when it could just be stored in a format which is queryable.

If it was instead represented as milliseconds (not a string) then the value would be easily queryable and now you can start to ask questions like "find me all of the users that took 5 minutes or more (duration >= 300000 milliseconds) to complete X simulation". With ISO 8601 it requires more work to be able to ask that question due to the lack of the support that I've found in databases such as MySQL.

You could attempt to ask my same question above with an ISO 8601 duration string, but would be incorrect in doing so as that compares the ASCII values, not what they actually mean when interpreted. Sure there are libraries & languages outside of the database which could handle this comparison, but have to do that to process data unnecessarily burdens the processing of statements so that a larger result set must be returned from the database and then made sense of instead of doing it in the database.

**Comment by fugu13** [comment]
ISO 8601 for durations is not going away, so it will always be a concern while querying. It has more expressive power than straight milliseconds, and has seen wide adoption. If you want that query to not involve parsing ISO 8601 durations in your database, I recommend parsing the duration in advance and then converting to an (approximate) number of milliseconds and inserting that in the database. Alternatively, do parse the durations in your database, but do so by creating a user defined function (perhaps named duration_in_milliseconds) and query using that. In either case the overhead is not large.

**Comment by canweriotnow** [comment]
I'm curious about the implications of using duration strings to represent
millisecond (or sub-millisecond) precision values, when an integer
representation would provide greater precision than a value that could be
interpreted as floating-point and therefore subject to rounding errors.
This might be an edge case (for now) but I certainly expect applications of
xAPI that will demand sub-millisecond precision and using an intermediate
decimal representation presents challenges, to say the least.

I suppose it could be fixed by using an extension to present the duration
as an integer of defined precision and then also using the spec-defined ISO
8601 duration for the sake of general conformance, with the expectation
that systems consuming the information will know to trust the former and
eschew the latter.

**Comment by fugu13** [comment]
A duration string can capture absolutely any precision that can be captured by an integer, though LRSs are allowed to truncate to deciseconds. Not all LRSs will (we don't, for example). Just don't convert it to a representation that loses precision on the way to your analysis (though at scales where sub-millisecond precision matters the level of error in a floating point representation is probably below the threshold measurement is feasible, so this is not a very practical thing to worry about).

To sum up: use an LRS that doesn't truncate to deciseconds and the current ISO 8601 string can hold whatever precision you want.

**Comment by DavidTPate** [comment]
Yeah, decimal representation of such things has the potential to lead to some really odd floating point edge cases.

The problem with using an extension to record the integer representation of the duration is that it is not interoperable. I say this because I could create statements with the extension `https://somesite.com/duration` and another provider could create a statement with the extension `https://someothersite/durationInMilliseconds`. Having ether the possibility of `duration` being able to be represented in an alternative format or a different property would be ideal in order to provide a standard other than extensions for interoperability.

That still leaves the potential that someone might not fill out the property, but the same issue exists with the current `duration` and it would at least provide the implication of interoperability since it would be based upon the specification.

**Comment by garemoko** [comment]
@DavidTPate If you create an extension and register at http://registry.tincanapi.com it's much less likely that others will create a new extension.

The big advantage of extensions is that you can start using them today, whereas an additional property would need to be part of a version 1.1.0 of the spec which is at minimum a year out from now, probably longer. 

Also, if you do create an extension and it becomes well used, then that becomes a good argument that the new property is needed. I don't know whether at that point we would canonize the extension or create a new property, but that's something to discuss later!

**Comment by andyjohnson** [comment]
Kept open for discussion, but it seems to have ceased for now, so closing.

**Comment by garemoko** [comment]
+1 to closing this (I know you've already closed it, but I checked so i'm recording that I checked and agree)
